### PR TITLE
DWARF: Update Ranges

### PR DESCRIPTION
**Pull request by kripken**
Pretty straightforward given all we have so far.

Note that fannkuch3_manyopts has an example of
a sequence of ranges of which some must be skipped
while others must not, showing we handle that by
skipping the bad ones and updating the remaining.

### CODE REVIEW

**Comment by dschuff** [comment]
what do you mean by ranges that must be skipped?

File: src/wasm/wasm-debug.cpp [hunk 1]
```diff
@@ -764,6 +766,55 @@ static void updateCompileUnits(const BinaryenDWARFInfo& info,
     });
 }
 
+static void updateRanges(llvm::DWARFYAML::Data& yaml,
+                         const LocationUpdater& locationUpdater) {
+  // In each range section, try to update the start and end. If we no longer
+  // have something to map them too, we must skip that part.
```
Comment (dschuff) [inline src/wasm/wasm-debug.cpp]: s/too/to

**Comment by kripken** [comment]
Imagine that we have a sequence of two `(begin, end)` spans
```
[(10, 20),
 (30, 40)]
```
It's possible `(10, 20)` maps in the new binary to `(110, 120)`
while `(30, 40)` was eliminated by the optimizer and we have
nothing valid to map it to. In that case we emit
```
[(110, 120)]
```

**Comment by dschuff** [comment]
Ah, got it. I guess if just an internal subrange were eliminated it would just work. what about if the beginning or end subrange is gone? I guess a downside of using real compiler output as test input is that we can't handpick edge cases.

**Comment by kripken** [comment]
Not sure what you mean by "beginning subrange" - is that `(10, 20)` in that example? It currently handles all subranges in the same way.

File: src/wasm/wasm-debug.cpp [hunk 2]
```diff
@@ -764,6 +766,55 @@ static void updateCompileUnits(const BinaryenDWARFInfo& info,
...
+  size_t skip = 0;
+  for (size_t i = 0; i < yaml.Ranges.size(); i++) {
+    auto& range = yaml.Ranges[i];
+    BinaryLocation oldStart = range.Start, oldEnd = range.End, newStart = 0,
+                   newEnd = 0;
+    // If this was not an end marker, try to find what it should be updated to.
+    if (oldStart != 0 && oldEnd != 0) {
+      if (locationUpdater.hasOldExprAddr(oldStart)) {
```
Comment (dschuff) [inline src/wasm/wasm-debug.cpp]: should we have some kind of combined lookup like `hasOldExprOrFuncStartAddr()` ? I guess when updating ranges we have no way to tell the context of the use...

File: src/wasm/wasm-debug.cpp [hunk 3]
```diff
@@ -764,6 +766,55 @@ static void updateCompileUnits(const BinaryenDWARFInfo& info,
...
+        newStart = locationUpdater.getNewExprAddr(oldStart);
+      } else if (locationUpdater.hasOldFuncStartAddr(oldStart)) {
+        newStart = locationUpdater.getNewFuncStartAddr(oldStart);
+      }
+      if (locationUpdater.hasOldExprEndAddr(oldEnd)) {
+        newEnd = locationUpdater.getNewExprEndAddr(oldEnd);
+      } else if (locationUpdater.hasOldFuncEndAddr(oldEnd)) {
+        newEnd = locationUpdater.getNewFuncEndAddr(oldEnd);
+      }
+      if (newStart == 0 || newEnd == 0) {
```
Comment (dschuff) [inline src/wasm/wasm-debug.cpp]: I guess in answer to my earlier question, If we have a range that represents a run of several instructions and we eliminate the last, I guess we don't have a good way to do "nearby" lookups in our current data structure.

**Comment by dschuff** [comment]
by "beginning subrange" I meant e.g. [10, 15)

**Comment by kripken** [comment]
I see now, thanks. Yeah, we could try to split subranges like `(10, 20)` if we know part of the instructions there. One issue is if we split into more than one piece then we need to rewrite all range section indexes. But what worries me more is that I'm not sure it would be correct or not - does part of a subrange make sense in all contexts? I have no idea...

File: src/wasm/wasm-debug.cpp (hunk 2 above)
Comment (kripken) [reply src/wasm/wasm-debug.cpp]: I could refactor a function for this, but I don't think that would be reusable anywhere else. This is the one place where we update both sides of a span (start, end) at once.

Comment (kripken) [reply src/wasm/wasm-debug.cpp]: For the second part, yeah, we can't tell the context atm. If the context would help I could work on getting it. However, I think it should be safe as it is, as we do have the context of whether we are at the start or end of a span, and the risky thing is just the ambiguity of the end of one thing which is equal to the start of the next, which we avoid here.

**Comment by dschuff** [comment]
Yeah, that might be the point where it makes more sense to just take the LLVM-style metadata approach (i.e. the debug info symbolically refers to IR constructs like functions and expressions) and regenerate it from scratch.
